### PR TITLE
BE-12-add-quiz-workflow

### PR DESCRIPTION
**Pull request by Randayy**

### CODE REVIEW

File: app/db/user_models.py [hunk 1]
```diff
@@ -111,3 +111,30 @@ class Answer(BaseTable):
...
         'questions.id', ondelete='CASCADE'), nullable=False)
     question = relationship('Question', back_populates='answers')
+
+
+class Result(BaseTable):
+    __tablename__ = 'results'
+
+    quiz_id = Column(UUID(as_uuid=True), ForeignKey(
+        'quizzes.id', ondelete='CASCADE'), nullable=False)
+    user_id = Column(UUID(as_uuid=True), ForeignKey(
```
Comment (Artemooon) [inline app/db/user_models.py:121]: use member_id here, only members can take a quiz

File: app/repositories/quiz_repository.py [hunk 2]
```diff
@@ -212,3 +213,109 @@ async def list_quizzes(self, company_id: UUID, page: int, limit: int):
...
+            user_id=current_user_id,
+            score=score,
+            created_at=datetime.utcnow()
+        )
+
+        self.db.add(quiz_result)
+        await self.db.commit()
+        await self.db.refresh(quiz_result)
+
+        for answer in user_answers:
```
Comment (Artemooon) [inline app/repositories/quiz_repository.py:251]: move this code to new function,  record_user_answers, this function is too large

File: app/repositories/quiz_repository.py [hunk 3]
```diff
@@ -212,3 +213,109 @@ async def list_quizzes(self, company_id: UUID, page: int, limit: int):
...
+            raise HTTPException(
+                status_code=403, detail="You are not a member of this company")
+
+# be-12
+    async def check_if_user_alredy_submitted_quiz(self, current_user_id: UUID, quiz_id: UUID):
+        result = await self.db.execute(select(Result).where(Result.quiz_id == quiz_id).where(Result.user_id == current_user_id))
+        result = result.scalars().first()
+        return result
+
+    async def submit_quiz_answers(self, user_answers: dict, current_user_id: UUID, quiz_id: UUID):
```
Comment (Artemooon) [inline app/repositories/quiz_repository.py]: I would move this function to the service, and leave only queries to the DB in repository

File: app/repositories/quiz_repository.py [hunk 4]
```diff
@@ -212,3 +213,109 @@ async def list_quizzes(self, company_id: UUID, page: int, limit: int):
...
+            raise HTTPException(
+                status_code=404, detail="You have not submitted any quiz in this company yet")
+
+        sum_of_scores = 0
+        quizzes = 0
+        for result in results:
+            sum_of_scores += result.score
+            quizzes += 1
+
+        avarage_mark = sum_of_scores/quizzes
```
Comment (Artemooon) [inline app/repositories/quiz_repository.py]: this logic also should be in the service

File: app/repositories/quiz_repository.py [hunk 5]
```diff
@@ -212,3 +213,109 @@ async def list_quizzes(self, company_id: UUID, page: int, limit: int):
...
+        result = await self.db.execute(select(Quiz).where(Quiz.id == quiz_id))
+        quiz = result.scalars().first()
+        if not quiz:
+            raise HTTPException(status_code=404, detail="Quiz not found")
+        questions = await self.db.execute(select(Question).where(Question.quiz_id == quiz_id))
+        questions = questions.scalars().all()
+        if not questions:
+            raise HTTPException(status_code=404, detail="Questions not found")
+        question_ids = [question.id for question in questions]
+        questions_answers_dict = []
```
Comment (Artemooon) [inline app/repositories/quiz_repository.py]: questions_answers_dict, but it's a list

Comment (Randayy) [reply app/repositories/quiz_repository.py]: Sorry it is list of dicts and I forgot to rename this

File: app/repositories/quiz_repository.py [hunk 6]
```diff
@@ -212,3 +213,109 @@ async def list_quizzes(self, company_id: UUID, page: int, limit: int):
...
+            question_answers = await self.get_question_answers(question_id)
+            questions_answers_dict.append(question_answers)
+        user_answers = user_answers['questions_answers']
+        user_answers_list_ids = [ans['answer_id']
+                                 for answer in user_answers for ans in answer['answers']]
+        correct_answers = 0
+        questions = 0
+        for question_answers in questions_answers_dict:
+            for question_answer in question_answers:
+                if question_answer['is_correct'] == True and question_answer['id'] in user_answers_list_ids:
```
Comment (Artemooon) [inline app/repositories/quiz_repository.py]: juist leave question_answer['is_correct'] and question_answer['id'] in user_answers_list_ids

Comment (Randayy) [reply app/repositories/quiz_repository.py]: fixed

File: app/repositories/quiz_repository.py (hunk 5 above)
Comment (Randayy) [reply app/repositories/quiz_repository.py]: renamed

File: app/repositories/quiz_repository.py (hunk 3 above)
Comment (Randayy) [reply app/repositories/quiz_repository.py]: refactored this huge function , made a few smaller and moved not queries to services

File: app/repositories/quiz_repository.py (hunk 2 above)
Comment (Randayy) [reply app/repositories/quiz_repository.py:251]: made new function save_user_answers

File: app/repositories/quiz_repository.py (hunk 4 above)
Comment (Randayy) [reply app/repositories/quiz_repository.py]: moved

File: app/db/user_models.py (hunk 1 above)
Comment (Randayy) [reply app/db/user_models.py:121]: I already check if this user is member of that company 
image

**Review by Artemooon — APPROVED** [review]

**Pull request merged**